src/hti/re_dash.cljd

```diff
@@ -323,28 +324,63 @@
 
 ;; Events Queue
 
-(defn- do-dispatch
-  [event-vec]
-  (let [event-id (first event-vec)
-        {:keys [interceptors event-type] :as registered-event} (event-id @events)
-        all-interceptors (concat [do-fx-interceptor
-                                  (prime-cofx-interceptor event-vec)]
-                                 interceptors)]
+(defonce ^:private ^coll/Queue events-queue
+  (coll/Queue))
 
-    (when-not registered-event
-      (throw (Exception. (str "Event not found: " event-id))))
+(defonce ^:private processing? (atom false))
 
-    (->> (await (interceptors-sweep-before event-type event-vec all-interceptors))
-         (interceptors-sweep-after all-interceptors))))
+(defn- dequeue
+  "If the queue is not empty, take the first event off the queue"
+  []
+  (when (.-isNotEmpty events-queue)
+    (.removeFirst events-queue)))
 
-(defonce ^:private events-queue
-  (let [controller (da/StreamController)]
+(defn- do-dispatch
+  "Process this event, then the next in the queue, until the queue is empty."
+  [event-vec]
 
-    (.listen (.-stream controller)
-             (fn [event-vec]
-               (await (do-dispatch event-vec))))
+  ;; Ensure we do not allow concurrent event execution.
+  ;; We open the gate again later after either
+  ;; - All events on the queue have been processed
+  ;; - Or an exception occurred
+  (reset! processing? true)
+
+  (try
+    (let [event-id                                               (first event-vec)
+          {:keys [interceptors event-type] :as registered-event} (event-id @events)
+          all-interceptors                                       (concat [do-fx-interceptor
+                                    (prime-cofx-interceptor event-vec)]
+                                   interceptors)]
+
+      (when-not registered-event
+        (throw (Exception. (str "Event not found: " event-id))))
+
+      (->> (interceptors-sweep-before event-type event-vec all-interceptors)
+           await
+           (interceptors-sweep-after all-interceptors)
+           await)
+
+      (if-let [next-event-vec (dequeue)]
+        (do-dispatch next-event-vec)
+        (reset! processing? false)))
+
+    (catch dynamic e
+      (reset! processing? false)
+      (throw e))))
```

Comment:
After an exception is thrown in an event handler, processing is stopped until another event is newly enqueued? Is this intended behavior?

Reply:
Not intended, more a consequence of not wanting to lose the exception.

You raise a valid point - why stop the current queue processing loop, yet still allow new events to be triggered.

I've refactored, so it logs the exception instead, but still continue to process the remainder of the queue in the current processing loop.

Reply:
Smooth! I should have this in production by tomorrow <3